src/connectors/hits-per-page/connectHitsPerPage.ts

```diff
@@ -4,101 +4,120 @@ import {
- * @example
- * // custom `renderFn` to render the custom HitsPerPage widget
- * function renderFn(HitsPerPageRenderingOptions, isFirstRendering) {
- *   var containerNode = HitsPerPageRenderingOptions.widgetParams.containerNode
- *   var items = HitsPerPageRenderingOptions.items
- *   var refine = HitsPerPageRenderingOptions.refine
- *
- *   if (isFirstRendering) {
- *     var markup = '<select></select>';
- *     containerNode.append(markup);
- *   }
- *
- *   const itemsHTML = items.map(({value, label, isRefined}) => `
- *     <option
- *       value="${value}"
- *       ${isRefined ? 'selected' : ''}
- *     >
- *       ${label}
- *     </option>
- *   `);
- *
- *   containerNode
- *     .find('select')
- *     .html(itemsHTML);
- *
- *   containerNode
- *     .find('select')
- *     .off('change')
- *     .on('change', e => { refine(e.target.value); });
- * }
- *
- * // connect `renderFn` to HitsPerPage logic
- * var customHitsPerPage = instantsearch.connectors.connectHitsPerPage(renderFn);
- *
- * // mount widget on the page
- * search.addWidgets([
- *   customHitsPerPage({
- *     containerNode: $('#custom-hits-per-page-container'),
- *     items: [
- *       {value: 6, label: '6 per page', default: true},
- *       {value: 12, label: '12 per page'},
- *       {value: 24, label: '24 per page'},
- *     ],
- *   })
- * ]);
- */
-export default function connectHitsPerPage(renderFn, unmountFn = noop) {
+export type HitsPerPageRendererOptionsItem = {
+  /**
+   * Label to display in the option.
+   */
+  label: string;
+
+  /**
+   * Number of hits to display per page.
+   */
+  value: number;
+
+  /**
+   * Indicates if it's the current refined value.
+   */
+  isRefined: boolean;
+};
+
+export type HitsPerPageConnectorParamsItem = {
+  /**
+   * Label to display in the option.
+   */
+  label: string;
+
+  /**
+   * Number of hits to display per page.
+   */
+  value: number;
+
+  /**
+   * The default hits per page on first search.
+   *
+   * @default false
+   */
+  default?: boolean;
+};
+
+export type HitsPerPageConnectorParams = {
+  /**
+   * Array of objects defining the different values and labels.
+   */
+  items: HitsPerPageConnectorParamsItem[];
+
+  /**
+   * Function to transform the items passed to the templates.
+   */
+  transformItems?: (
+    objects: HitsPerPageConnectorParamsItem[]
+  ) => HitsPerPageConnectorParamsItem[];
+};
+
+export type HitsPerPageRendererOptions<
+  THitsPerPageWidgetParams extends HitsPerPageConnectorParams = HitsPerPageConnectorParams
+> = {
+  /**
+   * Array of objects defining the different values and labels.
+   */
+  items: HitsPerPageRendererOptionsItem[];
+
+  /**
+   * Creates the URL for a single item name in the list.
+   */
+  createURL: (value: HitsPerPageRendererOptionsItem['value']) => string;
+
+  /**
+   * Sets the number of hits per page and trigger a search.
+   */
+  refine: (number: number) => any;
```

Comment:
I though we had an ESLint rule that forbids the use of argument name similar to a type name. I'd be in favor of removing this confusion.

```suggestion
  refine: (value: number) => any;
```